Modules/_posixsubprocess.c

```diff
@@ -383,6 +389,43 @@ _close_open_fds_maybe_unsafe(long start_fd, PyObject* py_fds_to_keep)
 #endif  /* else NOT (defined(__linux__) && defined(HAVE_SYS_SYSCALL_H)) */
 
 
+#ifdef VFORK_USABLE
+/* Reset dispositions for all signals to SIG_DFL except for ignored
+ * signals. This way we ensure that no signal handlers can run
+ * after we unblock signals in a child created by vfork().
+ */
+static void
+reset_signal_handlers(void)
+{
+    struct sigaction sa_dfl = {.sa_handler = SIG_DFL};
+    for (int sig = 1; sig < _NSIG; sig++) {
+        /* Dispositions for SIGKILL and SIGSTOP can't be changed. */
+        if (sig == SIGKILL || sig == SIGSTOP) {
+            continue;
+        }
+
+        struct sigaction sa;
+        /* C libraries usually return EINVAL for signals used
+         * internally (e.g. for thread cancellation), so simply
+         * skip errors here. */
+        if (sigaction(sig, NULL, &sa) == -1) {
+            continue;
```

Comment:
Here we have a problem that we need about `_NSIG` (usually 64 or 128) syscalls  to reset signal handlers. Glibc does exactly the same in `posix_spawn`, so no regression compared to glibc, but `musl` implements a kind of caching which avoids extra syscalls but is impossible to implement outside of a C library (without wrapping standard C functions).

One way to optimize this is to parse `SigCgt` field from `/proc/self/status`.

Reply:
I agree, I also don't want to optimize in the initial version, at least not with `/proc/self/status`. I've also discovered that `SigCgt` is incorrect (always zero) in Windows Subsystem for Linux (tested on Windows 10 1803).